svix-cli/src/relay/mod.rs

```diff
@@ -0,0 +1,467 @@
+                    },
+                })?
+                .into(),
+            )),
+        )
+        .await
+        {
+            Ok(Ok(_)) => { /* nothing to do  */ }
+            // The outer Result is for the timeout, the inner is if there was some other failure during `send`.
+            Ok(Err(_)) | Err(_) => {
+                anyhow::bail!("failed to complete handshake with Webhook Relay server: remote didn't accept start message");
+            }
+        }
+
+        // The assumption is the very first message we get from the websocket reader will be the
+        // response to our `MessageOut::Start` but it could also be any number of control messages.
+        // Keep reading until we see a `MessageIn::Start` or give up after some attempts.
+        const MAX_ATTEMPTS: u8 = 10;
+        let mut attempts = 0;
+        let start_response = loop {
+            if attempts > MAX_ATTEMPTS {
+                anyhow::bail!("failed to complete handshake with Webhook Relay server: no response from remote");
+            }
+            attempts += 1;
+
+            match tokio::time::timeout(SERVER_PING_PERIOD, ws_rx.next()).await {
+                Err(_timeout) => continue,
+                Ok(None) => {
+                    anyhow::bail!("no response from server for start message");
+                }
+                Ok(Some(msg)) => {
+                    let data = match msg? {
+                        // Control messages.
+                        Message::Close(Some(CloseFrame { code, reason }))
+                            if code == Policy && reason == SOCKET_IN_USE_REASON =>
+                        {
+                            return Err(TokenInUse.into())
+                        }
+                        Message::Close(_) => {
+                            anyhow::bail!("Relay server refused connection");
+                        }
+                        Message::Ping(_) | Message::Pong(_) | Message::Frame(_) => continue,
+
+                        // Messages that carry data we care to process.
+                        Message::Text(s) => s.into(),
+                        Message::Binary(bytes) => bytes,
+                    };
+
+                    match serde_json::from_slice::<MessageIn>(&data)? {
+                        // This is what we're waiting to see. A `MessageOut::Start` sent to the writer
+                        // should result in a `MessageInStart` coming back on the reader.
+                        MessageIn::Start { data, .. } => break data,
+                        MessageIn::Event { .. } => continue,
+                    };
+                }
+            }
+        };
+
+        if show_welcome_message {
+            println!(
+                r#"
+Webhook Relay is now listening at:
+{}
+
+All requests on this endpoint will be forwarded to your local URL:
+{}
+"#,
+                receive_url(&start_response.token),
+                self.local_url,
+            );
```

Comment:
Could use [`printdoc!`](https://docs.rs/indoc/latest/indoc/macro.printdoc.html) here and below.